gibbon-compiler/src/Gibbon/Passes/OptimizeADTLayout.hs

```diff
@@ -246,43 +249,116 @@ globallyOptimizeDataConLayout useGreedy prg@Prog{ddefs, fundefs, mainExp} = do
+
+  let greedyMain = 
+            (\(fList, dcon, edges) pr@Prog{ddefs=dd, fundefs=fds, mainExp=mexp} -> do 
+                              newSymDcon <- gensym (toVar dcon)
+                              let vals = P.map (\f -> (f, (dcon, newSymDcon), edges)) fList 
+                              P.foldrM greedyOptimization pr vals
+            )
+  if useGreedy then P.foldrM greedyMain prg funsToOptimizeTripleGreedy else P.foldrM solverMain prg funsToOptimizeTripleSolver
+
+
+isStrongConstr :: Constr -> Bool 
+isStrongConstr constr = case constr of 
+                              StrongConstr _ -> True 
+                              _ -> False
+
+-- Fold over all constraints to make sure we are not using duplicate constraints across different functions with possibly 
+-- different field meta data. Union all the meta data about a field globally by merding constraints.
+mergeConstraints :: [Constr] -> [Constr]
+mergeConstraints lst = case lst of 
+  [] -> [] 
+  [x] -> [x]
+  x@(WeakConstr e@(((a, ma), (b, mb)), _)):y -> let (x', y') = P.foldr (\constr (newConstr, tail) -> case constr of 
+                                                                                          WeakConstr e'@(((a', ma'), (b', mb')), wt) -> if (a, b) == (a', b') 
+                                                                                                                                        then
+                                                                                                                                          if isStrongConstr constr 
+                                                                                                                                          then (StrongConstr (((a', removeDuplicates $ ma' ++ ma), (b', removeDuplicates $ mb' ++ mb)), wt), tail) 
+                                                                                                                                          else (WeakConstr (((a', removeDuplicates $ ma' ++ ma), (b', removeDuplicates $ mb' ++ mb)), wt), tail) 
+                                                                                                                                        else (newConstr, tail ++ [constr])
+                                                                                          StrongConstr e'@(((a', ma'), (b', mb')), wt) -> if (a, b) == (a', b') 
+                                                                                                                                          then (StrongConstr (((a', removeDuplicates $ ma' ++ ma), (b', removeDuplicates $ mb' ++ mb)), wt), tail) 
+                                                                                                                                          else (newConstr, tail ++ [constr])
+
+
+                                                                        ) (x, []) y 
+                                                  in [x'] ++ (mergeConstraints y')
```

Comment:
Here and in many other places your indentation goes way beyond any reasonable limit. What is a reasonable limit? Most people try to not go beyond the column 100, but I prefer 80. Some of the lines here only start from column 138 and go as far as column 242. I think this should be fixed.

There are two reasons for this excessive indentation.

First, you still learn how to indent Haskell code in general. I encourage you to go to https://fourmolu.github.io/ and copy-paste any of your functions that go beyond 100 and see what it suggests. Most of the time it gives you an improvement. For example, a portion of this function that currently goes as far as  242 fourmolu reformats to stay in 148 columns.

We'd better have a convention enforced by CI to auto-format the code base, but this would be a big decision requiring a lot of thought. I'd suggest to not waiting for that and simply starting using a modern formatter (like fourmolu) locally regularly.

 Your second issue is too much nesting. When you see that the code approaches column 100 (using a reasonable formatter like fourmolu), you should seriously consider splitting your code into more functions.

---

Unrelated, this code has a bunch of duplications of `removeDuplicates $ mb' ++ mb` (similarly for `ma'`, `ma`). These common subexpressions could be `let`- (or `where-`) bound somewhere and used here.